presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlock.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.Type;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.spi.block.BlockUtil.intSaturatedCast;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class MapBlock
+        extends AbstractMapBlock
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapBlock.class).instanceSize();
+
+    private final int startOffset;
+    private final int positionCount;
+
+    private final boolean[] valueIsNull;
+    private final int[] offsets;
+    private final Block keyBlock;
+    private final Block valueBlock;
+    private final int[] hashTable; // hash to location in map;
+
+    private int sizeInBytes;
+    private final int retainedSizeInBytes;
+
+    /**
+     * @param keyNativeHashCode (T)long
+     * @param keyBlockNativeEquals (T, Block,int)long
```

Comment:
We don't generally validate type of MethodHandles. And I think the value it provides is pretty minimal. I can certainly add it if you feel strongly.